clippy_lints/src/mut_key.rs

```diff
@@ -19,10 +19,14 @@ declare_clippy_lint! {
     /// so having types with interior mutability is a bad idea.
     ///
     /// ### Known problems
-    /// It's correct to use a struct, that contains interior mutability
-    /// as a key, when its `Hash` implementation doesn't access any of the interior mutable types.
-    /// However, this lint is unable to recognize this, so it causes a false positive in theses cases.
-    /// The `bytes` crate is a great example of this.
+    /// It's correct to use a struct that contains interior mutability as a key, when its
+    /// implementation of `Hash` or `Ord` doesn't access any of the interior mutable types.
+    /// However, this lint is unable to recognize this, so it will often cause false positives in
+    /// theses cases.  The `bytes` crate is a great example of this.
+    ///
+    /// Additionally, if a struct/enum contains interior mutability behind indirection then this
+    /// lint is unlikely to catch it.  For example, `struct BadKey<'a>(&'a Cell<usize>)` will not
+    /// be detected by this lint.
     ///
     /// ### Example
     /// ```rust
```

Comment:
Huh isn't this example an instance of the false negative described above? Could you update that while you're at it? A simple `Cell` example would be good. Also feel free to update lint docs and description to be more precise (use "interior mutability").

Reply:
Huh?  The first example is a false positive, so this false negative can't be an instance of it.  I've updated the docs though to hopefully make for easier reading :smile:.

Reply:
I've rebased the changes into the same 3 commit messages as before, so feel free to close this conversation if the changes are good :grin:.

Reply:
I was referring to the ###Example section.

Reply:
Ah I see!  The example looks fine to me - it's definitely on the verbose side, but it does correctly trigger the lint.  We could potentially remove the `impl` blocks and the example would still work.

Reply:
Oh it wasn't firing for me because it doesn't fire on `let _: .. = ..` 🤦 . Yeah if you could fix that and trim out the impls that would be good.

Reply:
> Oh it wasn't firing for me because it doesn't fire on `let _: .. = ..` facepalm .

Ah yes I got gotcha-ed by that exact same thing when writing the tests :sweat_smile:

>  Yeah if you could fix that and trim out the impls that would be good

Right I'll do this.  Although, thinking about it, I think the lint becomes a false positive if I remove the `impl` blocks.  We can't `#[derive(Hash, PartialEq, Eq)]` because `AtomicUsize` doesn't impl any of those 3 traits.  So I think the example is as short as it can be without being subtly wrong.  What d'ya reckon @camsteffen?  If you're OK with it, then feel free to resolve this conversation :).

Reply:
Makes sense